lib/read/lib/Cardano/Wallet/Read/Eras/EraFun.hs

```diff
@@ -194,42 +213,58 @@ newtype EraFunK src ft = EraFunK {fromEraFunK :: EraFun src (K ft)}
 instance Functor (EraFunK src) where
     fmap :: forall a b. (a -> b) -> EraFunK src a -> EraFunK src b
     fmap f (EraFunK g) =
-        EraFunK (toEraFun $ map_NP q $ fromEraFun g)
+        EraFunK (MkEraFun $ map_NP q $ fromEraFun g)
       where
         q :: (-.->) src (K a) era -> (-.->) src (K b) era
         q (Fn h) = Fn $ \x -> K . f $ unK $ h x
 
 instance Applicative (EraFunK src) where
-    pure x = EraFunK $ toEraFun $ pure_NP $ Fn $ \_ -> K x
+    pure x = EraFunK $ MkEraFun $ pure_NP $ Fn $ \_ -> K x
     EraFunK f <*> EraFunK g =
-        EraFunK $ toEraFun $ zipWith_NP q (fromEraFun f) (fromEraFun g)
+        EraFunK $ MkEraFun $ zipWith_NP q (fromEraFun f) (fromEraFun g)
       where
         q (Fn h) (Fn j) = Fn $ \src -> K $ unK (h src) $ unK $ j src
 
-type AllEraValue f = EraFun (K ()) f
+-- | A constant era 'EraFun' wrapped to expose the semigroup instance
+newtype AllEraValue f = AllEraValue {_unAllEraValue :: EraFun (K ()) f}
+
+-- | A pattern to construct/deconstruct an 'AllEraValue'
+pattern AllEraValueP
+    :: f ByronEra
+    -> f ShelleyEra
+    -> f AllegraEra
+    -> f MaryEra
+    -> f AlonzoEra
+    -> f BabbageEra
+    -> f ConwayEra
+    -> AllEraValue f
+pattern AllEraValueP
+    { byronVal
+    , shelleyVal
+    , allegraVal
+    , maryVal
+    , alonzoVal
+    , babbageVal
+    , conwayVal
+    } <-
+    AllEraValue
+        ( EraFun
+                { byronFun = (mkConst -> byronVal)
+                , shelleyFun = (mkConst -> shelleyVal)
+                , allegraFun = (mkConst -> allegraVal)
+                , maryFun = (mkConst -> maryVal)
+                , alonzoFun = (mkConst -> alonzoVal)
+                , babbageFun = (mkConst -> babbageVal)
+                , conwayFun = (mkConst -> conwayVal)
+                }
+            )
 
```

Comment:
```suggestion
  where
    AllEraValueP
        byronVal
        shelleyVal
        allegraVal
        maryVal
        alonzoVal
        babbageVal
        conwayVal
        = AllEraValue
            ( EraFun
                { byronFun = const byronVal
                , shelleyFun = const shelleyVal
                , allegraFun = const allegraVal
                , maryFun = const maryVal
                , alonzoFun = const alonzoVal
                , babbageFun = const babbageVal
                , conwayFun = const conwayVal
                }
            )
```

The comment promises "construct/deconstruct", so we should make the pattern bidirectional.

Either that, or change the comment to
```
-- | A pattern to deconstruct an 'AllEraValue'
```

Reply:
I think it's bidirectional, and I cannot make your suggestion to compile

Reply:
```haskell
-- | A pattern to construct/deconstruct an 'AllEraValue'
pattern AllEraValueP
    :: f ByronEra
    -> f ShelleyEra
    -> f AllegraEra
    -> f MaryEra
    -> f AlonzoEra
    -> f BabbageEra
    -> f ConwayEra
    -> AllEraValue f
pattern AllEraValueP
    { byronVal
    , shelleyVal
    , allegraVal
    , maryVal
    , alonzoVal
    , babbageVal
    , conwayVal
    } <-
    AllEraValue
        ( EraFun
                { byronFun = mkConst -> byronVal
                , shelleyFun = mkConst -> shelleyVal
                , allegraFun = mkConst -> allegraVal
                , maryFun = mkConst -> maryVal
                , alonzoFun = mkConst -> alonzoVal
                , babbageFun = mkConst -> babbageVal
                , conwayFun = mkConst -> conwayVal
                }
            )
    where
        AllEraValueP
            byronVal'
            shelleyVal'
            allegraVal'
            maryVal'
            alonzoVal'
            babbageVal'
            conwayVal' =
                AllEraValue
                    $ EraFun
                        { byronFun = const byronVal'
                        , shelleyFun = const shelleyVal'
                        , allegraFun = const allegraVal'
                        , maryFun = const maryVal'
                        , alonzoFun = const alonzoVal'
                        , babbageFun = const babbageVal'
                        , conwayFun = const conwayVal'
                        }
```
is working